### ISSUE TITLE
Allow implicit narrowing conversion if constexpr is known to fit

### ISSUE THREAD

**Issue by Niko-O** [foonathan/type_safe#127]
First of all, thank you for mitigating one of the two biggest problems (IMO) with C++!

I don't have a lot of experience with C++-isms, so I can't figure it out myself. Is there any way to tell C++ to allow assigning a value of a wider type to a variable of a narrower type, if the value is known to be small enough to fit? Like this:
`type_safe::types::uint8_t someSmallValue = 0;`
If possible also this:
```
constexpr unsigned short something = 0;
type_safe::types::uint8_t someSmallValue = something;
```
And this:
```
constexpr type_safe::types::uint16_t something = 0; // or 0_i16
type_safe::types::uint8_t someSmallValue = something;
```
Currently it is necerssary to put the "_i8" on the literal or use a narrow_cast, which can become a bit cumbersome:
`type_safe::types::uint8_t someSmallValue = 0_i8;`
`type_safe::types::uint8_t someSmallValue = type_safe::narrow_cast<type_safe::types::uint8_t>(something);`
And this won't work at all:
`type_safe::types::uint8_t someSmallValue = type_safe::narrow_cast<type_safe::types::uint8_t>(0);`

Also, could you maybe elaborate on why the integer class "intentionally does not provide the bitwise operations."?

Edit: And is there any way to make it less cumbersome to use these types as array indices? E.g. in `things[y.get()][x.get()];` it's always necessary to write `.get()` there.

**Comment by foonathan** [comment]
> Is there any way to tell C++ to allow assigning a value of a wider type to a variable of a narrower type, if the value is known to be small enough to fit?

Unfortunately not. You need to do `std::uint16_t(constant)`.

> Also, could you maybe elaborate on why the integer class "intentionally does not provide the bitwise operations."?

My philosophy is that there are integer classes to do arithmetic and bit containers to do bitwise operations. type_safe provides the former types.

> And is there any way to make it less cumbersome to use these types as array indices? E.g. in things[y.get()][x.get()]; it's always necessary to write .get() there.

Again, unfortunately not.

**Comment by Niko-O** [comment]
Thanks for the clarification. Bummer that those two things are not possible.
If you don't mind me asking some more questions here:

I integer.hpp, `integer_type value_;` is declared at the very end of the class. How does that work?

And how do I get from a wide raw type to narrow wrapped one?
For example:
```
uint8_t* start = ...; // Points to somewhere in an array
uint8_t* end = ...; // Points to somewhere later (+1) in the same array
auto rawLength = end - start; // Apparently this is of type long, which I tested to be 8 bytes wide
```
I now need the length as a `uint16_t`.  What is the **proper** way of doing that?
`uint16_t length = static_cast<unsigned short>(rawLength);` just truncates the value silently.
`uint16_t length = narrow_cast<uint16_t>(static_cast<int64_t>(rawLength));` works, but it's quite verbose and I have to match the original type, even though I don't really care about it.

I came up with this helper to make it less verbose:
```
template<typename TResult, typename TRaw>
TResult castHelper(TRaw raw)
{
    return narrow_cast<TResult>(static_cast<typename integer<TRaw>>(raw));
}
```
But do you have a specific way of doing this?

**Comment by foonathan** [comment]
> I integer.hpp, integer_type value_; is declared at the very end of the class. How does that work?

What do you mean?

> I now need the length as a uint16_t. What is the proper way of doing that?

`auto length = narrow_cast<uint16_t>(raw_length)` should work. What's the error?

**Comment by Niko-O** [comment]
> What do you mean?
Using something before declaring it is not something that usually works in C++.
> What's the error?
I have to confess, I added this constructor to the `integer` class: `TYPE_SAFE_FORCE_INLINE constexpr integer() : value_(0) { }` Everything else is as-is. Then I have this simple example in the only .cpp file of the project, so nothing should interfere:
```C++
#include <type_safe/config.hpp>
#include <type_safe/integer.hpp>
#include <type_safe/types.hpp>
#include <type_safe/narrow_cast.hpp>

int main()
{
    type_safe::types::uint8_t array[100];
    type_safe::types::uint8_t* start = &array[10];
    type_safe::types::uint8_t* end = &array[12];
    auto rawLength = end - start;
    type_safe::types::uint16_t length = type_safe::narrow_cast<type_safe::types::uint16_t>(rawLength);

    return 0;
}
```
I am using Qt Creator with C++ 17 and I am getting this error at the top of the code editor:
```
Warning: The code model could not parse an included file, which might lead to incorrect code completion and hightlighting, for example.
narrow_cast.hpp:47:29: error: cannot convert 'type_safe::integer<unsigned short>' to 'long' without a conversion operator
main.cpp:1:1: note: in file included from /media/ShareW/Repositories/Tutorials and Tests/TestCpp/main.cpp:1:
main.cpp:4:10: note: in file included from /media/ShareW/Repositories/Tutorials and Tests/TestCpp/main.cpp:4:
narrow_cast.hpp:74:20: note: in instantiation of function template specialization 'type_safe::detail::is_narrowing<type_safe::integer<unsigned short>, long>' requested here
main.cpp:12:52: note: in instantiation of function template specialization 'type_safe::narrow_cast<type_safe::integer<unsigned short>, long, void>' requested here
```
That narrow_cast.hpp:47:29 is this:
```C++
    template <typename Target, typename Source>
    TYPE_SAFE_FORCE_INLINE constexpr bool is_narrowing(
        const Source& source,
        typename std::enable_if<std::is_integral<Source>::value, int>::type = 0)
    {
        using limits = std::numeric_limits<Target>;
        return sizeof(Target) < sizeof(Source) // no narrowing possible
               && (source > Source(limits::max()) // Line 47, column 29 is the start of "Source"
                   || source < Source(limits::min())); // otherwise check bounds
    }
```
So assuming `Target` is `type_safe::types::uint16_t` and `Source` is `long`, this tries to "function-style cast" `std::numeric_limits<type_safe::types::uint16_t>::max()`, which is itself a `type_safe::types::uint16_t`, to `long`.

**Comment by foonathan** [comment]
I've fixed this one in 1c0d3a6, sorry for the delay. Note that a workaround would have been `narrow_cast<uint16_t>` instead.
